src/NamingStrategy/MapNamingStrategy.php

```diff
@@ -9,83 +9,100 @@
 {
     /**
-     * Map for hydrate name conversion.
-     *
      * @var string[]
      */
-    protected $mapping = [];
+    private $extractionMap = [];
 
     /**
-     * Reversed map for extract name conversion.
-     *
      * @var string[]
      */
-    protected $reverse = [];
+    private $hydrationMap = [];
 
     /**
-     * @param string[]      $mapping Map for name conversion on hydration
-     * @param null|string[] $reverse Reverse map for name conversion on extraction
+     * @param null|string[] $hydrationMap A map of string keys and values for
+     *     translation of hydrated field names. If not provided, the result of
+     *     an array_flip($extractionMap) will be used.
+     * @param null|string[] $extractionMap A map of string keys and values for
+     *     translation of extracted field names. If not provided, the result of
+     *     an array_flip($hydrationMap) will be used.
+     * @throws Exception\InvalidArgumentException if neither $hydrationMap nor
+     *     $extractionMap are provided.
+     * @throws Exception\InvalidArgumentException when flipping either the
+     *     $hydrationMap or $extractionMap, and any value is a non-string,
+     *     non-int value.
      */
-    public function __construct(array $mapping, ?array $reverse = null)
+    public function __construct(?array $hydrationMap = null, ?array $extractionMap = null)
     {
-        $this->mapping = $mapping;
-        $this->reverse = $reverse ?: $this->flipMapping($mapping);
+        if (null === $hydrationMap && null === $extractionMap) {
+            throw new Exception\InvalidArgumentException(sprintf(
+                '%s requires one or both of an array $hydrationMap and array $extractionMap;'
+                . ' neither provided',
+                __CLASS__
+            ));
+        }
+
+        if (null === $extractionMap && is_array($hydrationMap)) {
+            $extractionMap = $this->flipMapping($hydrationMap);
+        }
+
+        if (null === $hydrationMap && is_array($extractionMap)) {
+            $hydrationMap = $this->flipMapping($extractionMap);
+        }
+
+        $this->extractionMap = (array) $extractionMap;
+        $this->hydrationMap  = (array) $hydrationMap;
     }
 
     /**
-     * Safely flip mapping array.
-     *
-     * @param  string[] $array Array to flip
-     * @return string[] Flipped array
-     * @throws InvalidArgumentException
+     * {@inheritDoc}
      */
-    protected function flipMapping(array $array) : array
+    public function extract(string $name, ?object $object = null) : string
     {
-        array_walk($array, function ($value) {
-            if (! is_string($value) && ! is_int($value)) {
-                throw new InvalidArgumentException('Mapping array can\'t be flipped because of invalid value');
-            }
-        });
-
-        return array_flip($array);
+        return array_key_exists($name, $this->extractionMap)
```

Comment:
Since we know that all `$this->extractionMap` is `array<string, string>`, this method can be simplified to:

```php
return $this->extractionMap[$name] ?? $name;
```